### PR TITLE
use PMacc assert and verify

### PR DESCRIPTION
**Pull request by psychocoderHPC**
substitute `assert` with `PMACC_ASSERT` or `PMACC_VERIFY`

Both calls are available since #1578

### CODE REVIEW

File: src/libPMacc/include/cuSTL/container/CartBuffer.tpp [hunk 1]
```diff
@@ -198,7 +199,7 @@ CartBuffer<Type, T_dim, Allocator, Copier, Assigner>::operator=
             std::stringstream() << "Assignment: Sizes of buffers do not match: "
                 << this->size() << " <-> " << rhs.size() << std::endl).str());
 #else
-    assert(rhs.size() == this->size());
+    PMACC_ASSERT(rhs.size() == this->size());
```
Comment (psychocoderHPC) [inline src/libPMacc/include/cuSTL/container/CartBuffer.tpp]: this create a WARNINGS: `warning: calling a __host__ function from a __host__ __device__ function is not allowed
`

File: src/libPMacc/include/cuSTL/algorithm/kernel/run-time/Foreach.hpp [hunk 2]
```diff
@@ -130,10 +130,10 @@ math::Size_t<DIM3> getBestCudaBlockDim(const math::Size_t<dim> gridDim)
...
             this->_blockDim = getBestCudaBlockDim(p_zone.size);                                                     \
                                                                                                                     \
-        assert(this->_blockDim.productOfComponents() <= cudaSpecs::maxNumThreadsPerBlock);                          \
-        assert(this->_blockDim.x() <= cudaSpecs::MaxNumThreadsPerBlockDim::x::value);                               \
-        assert(this->_blockDim.y() <= cudaSpecs::MaxNumThreadsPerBlockDim::y::value);                               \
-        assert(this->_blockDim.z() <= cudaSpecs::MaxNumThreadsPerBlockDim::z::value);                               \
+        PMACC_VERIFY(this->_blockDim.productOfComponents() <= cudaSpecs::maxNumThreadsPerBlock);                    \
+        PMACC_VERIFY(this->_blockDim.x() <= cudaSpecs::MaxNumThreadsPerBlockDim::x::value);                         \
+        PMACC_VERIFY(this->_blockDim.y() <= cudaSpecs::MaxNumThreadsPerBlockDim::y::value);                         \
+        PMACC_VERIFY(this->_blockDim.z() <= cudaSpecs::MaxNumThreadsPerBlockDim::z::value);                         \
```
Comment (PrometheusPi) [inline src/libPMacc/include/cuSTL/algorithm/kernel/run-time/Foreach.hpp]: please add spaces `PMACC_VERIFY( this->_blockDim.z() <= cudaSpecs::MaxNumThreadsPerBlockDim::z::value ); `

File: src/libPMacc/include/eventSystem/tasks/ITask.hpp [hunk 3]
```diff
@@ -54,7 +55,7 @@ namespace PMacc
             static id_t globalId = 1;
 
             myId = globalId++;
-            assert(myId > 0);
+            PMACC_ASSERT(myId > 0);
```
Comment (PrometheusPi) [inline src/libPMacc/include/eventSystem/tasks/ITask.hpp]: please add spaces `PMACC_ASSERT( myId > 0 );`

File: src/libPMacc/include/eventSystem/tasks/StreamTask.tpp [hunk 4]
```diff
@@ -26,7 +26,7 @@
 //#include "eventSystem/EventSystem.hpp"
 #include "eventSystem/tasks/StreamTask.hpp"
 #include "eventSystem/streams/EventStream.hpp"
-#include "Environment.hpp"
```
Comment (PrometheusPi) [inline src/libPMacc/include/eventSystem/tasks/StreamTask.tpp]: @psychocoderHPC why did you remove this include?

File: src/libPMacc/include/mappings/kernel/BorderMapping.hpp [hunk 5]
```diff
@@ -74,7 +75,7 @@ namespace PMacc
          */
         HINLINE BorderMapping(const BaseClass& base, PMacc::ExchangeType direction): BaseClass(base), m_direction(direction)
         {
-            assert(direction != 0);
+            PMACC_ASSERT(direction != 0);
```
Comment (PrometheusPi) [inline src/libPMacc/include/mappings/kernel/BorderMapping.hpp]: please add spaces

File: src/libPMacc/include/mappings/kernel/MappingDescription.hpp [hunk 6]
```diff
@@ -64,9 +65,9 @@ class MappingDescription
         {
             minBlock = std::min(minBlock, gridSuperCells[2]);
         }
-        assert((guardingSuperCells == 0) || (minBlock >= 3));
+        PMACC_VERIFY((guardingSuperCells == 0) || (minBlock >= 3));
```
Comment (PrometheusPi) [inline src/libPMacc/include/mappings/kernel/MappingDescription.hpp]: please add spaces

File: src/libPMacc/include/mappings/kernel/MappingDescription.hpp [hunk 7]
```diff
@@ -64,9 +65,9 @@ class MappingDescription
...
         /*border block count must smaller or equal to core blocks count*/
-        assert(borderSuperCells <= (minBlock - (2 * guardingSuperCells)));
+        PMACC_VERIFY(borderSuperCells <= (minBlock - (2 * guardingSuperCells)));
```
Comment (PrometheusPi) [inline src/libPMacc/include/mappings/kernel/MappingDescription.hpp]: please add spaces

File: src/libPMacc/include/memory/buffers/Buffer.hpp [hunk 8]
```diff
@@ -167,7 +167,7 @@ namespace PMacc
         virtual void setCurrentSize(const size_t newsize)
         {
             __startOperation(ITask::TASK_HOST);
-            assert(static_cast<size_t>(newsize) <= static_cast<size_t>(data_space.productOfComponents()));
+            PMACC_ASSERT(static_cast<size_t>(newsize) <= static_cast<size_t>(data_space.productOfComponents()));
```
Comment (PrometheusPi) [inline src/libPMacc/include/memory/buffers/Buffer.hpp]: please add spaces

File: src/libPMacc/include/memory/buffers/DeviceBufferIntern.hpp [hunk 9]
```diff
@@ -212,15 +211,15 @@ class DeviceBufferIntern : public DeviceBuffer<TYPE, DIM>
     void copyFrom(HostBuffer<TYPE, DIM>& other)
     {
 
-        assert(this->isMyDataSpaceGreaterThan(other.getCurrentDataSpace()));
+        PMACC_ASSERT(this->isMyDataSpaceGreaterThan(other.getCurrentDataSpace()));
```
Comment (PrometheusPi) [inline src/libPMacc/include/memory/buffers/DeviceBufferIntern.hpp]: please add spaces

File: src/libPMacc/include/memory/buffers/DeviceBufferIntern.hpp [hunk 10]
```diff
@@ -212,15 +211,15 @@ class DeviceBufferIntern : public DeviceBuffer<TYPE, DIM>
...
         Environment<>::get().Factory().createTaskCopyHostToDevice(other, *this);
 
     }
 
     void copyFrom(DeviceBuffer<TYPE, DIM>& other)
     {
 
-        assert(this->isMyDataSpaceGreaterThan(other.getCurrentDataSpace()));
+        PMACC_ASSERT(this->isMyDataSpaceGreaterThan(other.getCurrentDataSpace()));
```
Comment (PrometheusPi) [inline src/libPMacc/include/memory/buffers/DeviceBufferIntern.hpp]: please add spaces

File: src/libPMacc/include/memory/buffers/ExchangeIntern.hpp [hunk 11]
```diff
@@ -52,7 +51,7 @@ namespace PMacc
         Exchange<TYPE, DIM>(exchange, communicationTag), deviceDoubleBuffer(NULL)
         {
 
-            assert(!guardingCells.isOneDimensionGreaterThan(memoryLayout.getGuard()));
+            PMACC_ASSERT(!guardingCells.isOneDimensionGreaterThan(memoryLayout.getGuard()));
```
Comment (PrometheusPi) [inline src/libPMacc/include/memory/buffers/ExchangeIntern.hpp]: please add spaces

File: src/libPMacc/include/memory/buffers/HostBufferIntern.hpp [hunk 12]
```diff
@@ -93,7 +92,7 @@ class HostBufferIntern : public HostBuffer<TYPE, DIM>
 
     void copyFrom(DeviceBuffer<TYPE, DIM>& other)
     {
-        assert(this->isMyDataSpaceGreaterThan(other.getCurrentDataSpace()));
+        PMACC_ASSERT(this->isMyDataSpaceGreaterThan(other.getCurrentDataSpace()));
```
Comment (PrometheusPi) [inline src/libPMacc/include/memory/buffers/HostBufferIntern.hpp]: please add spaces

File: src/libPMacc/include/memory/buffers/MappedBufferIntern.hpp [hunk 13]
```diff
@@ -99,13 +98,13 @@ class MappedBufferIntern : public DeviceBuffer<TYPE, DIM>
 
     void copyFrom(HostBuffer<TYPE, DIM>& other)
     {
-        assert(this->isMyDataSpaceGreaterThan(other.getCurrentDataSpace()));
+        PMACC_ASSERT(this->isMyDataSpaceGreaterThan(other.getCurrentDataSpace()));
```
Comment (PrometheusPi) [inline src/libPMacc/include/memory/buffers/MappedBufferIntern.hpp]: please add spaces

File: src/libPMacc/include/memory/buffers/MappedBufferIntern.hpp [hunk 14]
```diff
@@ -99,13 +98,13 @@ class MappedBufferIntern : public DeviceBuffer<TYPE, DIM>
...
         Environment<>::get().Factory().createTaskCopyHostToDevice(other, *this);
     }
 
     void copyFrom(DeviceBuffer<TYPE, DIM>& other)
     {
-        assert(this->isMyDataSpaceGreaterThan(other.getCurrentDataSpace()));
+        PMACC_ASSERT(this->isMyDataSpaceGreaterThan(other.getCurrentDataSpace()));
```
Comment (PrometheusPi) [inline src/libPMacc/include/memory/buffers/MappedBufferIntern.hpp]: please add spaces

File: src/libPMacc/include/memory/buffers/MultiGridBuffer.hpp [hunk 15]
```diff
@@ -220,7 +221,7 @@ class MultiGridBuffer
 
     GridBuffer<Type, DIM>& getGridBuffer(typename BufferNames::Names name)
     {
-        assert(name >= 0 && name < BufferNames::Count);
+        PMACC_ASSERT(name >= 0 && name < BufferNames::Count);
```
Comment (PrometheusPi) [inline src/libPMacc/include/memory/buffers/MultiGridBuffer.hpp]: please add spaces

File: src/libPMacc/include/mpi/MPIReduce.hpp [hunk 16]
```diff
@@ -61,7 +59,7 @@ struct MPIReduce
     template<class MPIMethod>
     bool hasResult(const MPIMethod & method) const
     {
-        assert(isMPICommInitialized == true);
+        PMACC_ASSERT(isMPICommInitialized == true);
```
Comment (PrometheusPi) [inline src/libPMacc/include/mpi/MPIReduce.hpp]: please add spaces

File: src/libPMacc/include/particles/ParticlesBase.hpp [hunk 17]
```diff
@@ -179,7 +180,7 @@ class ParticlesBase : public SimulationFieldHelper<T_MappingDesc>
      */
     BufferType& getParticlesBuffer()
     {
-        assert(particlesBuffer != NULL);
+        PMACC_ASSERT(particlesBuffer != NULL);
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/ParticlesBase.hpp]: please add spaces

File: src/libPMacc/include/particles/memory/buffers/ParticlesBuffer.hpp [hunk 18]
```diff
@@ -330,7 +330,7 @@ class ParticlesBuffer
     DataSpace<DIM> getSuperCellsCount()
     {
 
-        assert(superCells != NULL);
+        PMACC_ASSERT(superCells != NULL);
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/memory/buffers/ParticlesBuffer.hpp]: please add spaces

File: src/libPMacc/include/particles/memory/buffers/ParticlesBuffer.hpp [hunk 19]
```diff
@@ -342,7 +342,7 @@ class ParticlesBuffer
     GridLayout<DIM> getSuperCellsLayout()
     {
 
-        assert(superCells != NULL);
+        PMACC_ASSERT(superCells != NULL);
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/memory/buffers/ParticlesBuffer.hpp]: please add spaces

File: src/libPMacc/include/particles/memory/buffers/StackExchangeBuffer.hpp [hunk 20]
```diff
@@ -91,8 +91,8 @@ namespace PMacc
          */
         ExchangePushDataBox<vint_t, FRAME, DIM> getDeviceExchangePushDataBox()
         {
-            assert(stack.getDeviceBuffer().hasCurrentSizeOnDevice() == true);
-            assert(stackIndexer.getDeviceBuffer().hasCurrentSizeOnDevice() == true);
+            PMACC_ASSERT(stack.getDeviceBuffer().hasCurrentSizeOnDevice() == true);
+            PMACC_ASSERT(stackIndexer.getDeviceBuffer().hasCurrentSizeOnDevice() == true);
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/memory/buffers/StackExchangeBuffer.hpp]: please add spaces

File: src/libPMacc/include/particles/tasks/TaskReceiveParticlesExchange.hpp [hunk 21]
```diff
@@ -83,7 +84,7 @@ namespace PMacc
                     if (NULL == Environment<>::get().Manager().getITaskIfNotFinished(tmpEvent.getTaskId()))
                     {
                         state=Wait;
-                        assert(lastSize <= maxSize);
+                        PMACC_ASSERT(lastSize <= maxSize);
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/tasks/TaskReceiveParticlesExchange.hpp]: please add spaces

File: src/libPMacc/include/particles/tasks/TaskSendParticlesExchange.hpp [hunk 22]
```diff
@@ -82,7 +83,7 @@ namespace PMacc
                 case WaitForSend:
                     if (NULL == Environment<>::get().Manager().getITaskIfNotFinished(tmpEvent.getTaskId()))
                     {
-                        assert(lastSize<=maxSize);
+                        PMACC_ASSERT(lastSize<=maxSize);
```
Comment (PrometheusPi) [inline src/libPMacc/include/particles/tasks/TaskSendParticlesExchange.hpp]: please add spaces - especially around `<=`

File: src/libPMacc/test/random/2DDistribution.cu [hunk 23]
```diff
@@ -188,7 +189,7 @@ void runTest(uint32_t numSamples)
             mean += PMacc::math::linearize(size.shrink<1>(1), idx) * static_cast<uint64_t>(val);
         }
     }
-    assert(totalNumSamples == uint64_t(rngSize.productOfComponents()) * uint64_t(numSamples));
+    PMACC_ASSERT(totalNumSamples == uint64_t(rngSize.productOfComponents()) * uint64_t(numSamples));
```
Comment (PrometheusPi) [inline src/libPMacc/test/random/2DDistribution.cu]: please add spaces

File: src/libPMacc/include/cuSTL/algorithm/kernel/run-time/Foreach.hpp (hunk 2 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/cuSTL/algorithm/kernel/run-time/Foreach.hpp]: I keeped the code style from the code around therefore there are no spaces

File: src/libPMacc/include/eventSystem/tasks/StreamTask.tpp (hunk 4 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/eventSystem/tasks/StreamTask.tpp]: already include in line 25

**Comment by psychocoderHPC** [comment]
I used the over the thump rule to choose verify/assert:
- if the checked value is influenced by the user (e.g. PIConGPU param, ...) than use **verify**
- Pointer checks use **assert**

File: src/libPMacc/include/memory/buffers/MultiGridBuffer.hpp (hunk 15 above)
Comment (PrometheusPi) [inline src/libPMacc/include/memory/buffers/MultiGridBuffer.hpp]: @psychocoderHPC Just for understanding: why do you throw an assert if name is in range [0, Count)?

**Comment by PrometheusPi** [comment]
@psychocoderHPC Thanks for the clarification.

File: src/libPMacc/include/memory/buffers/MultiGridBuffer.hpp (hunk 15 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/memory/buffers/MultiGridBuffer.hpp]: You god me one second where I thought  this is a bug but than ...
The assert is only fire an exception if the expression is evaluated to `false`. This means that only valid names are allowed.

**Comment by psychocoderHPC** [comment]
offline discussed: The primary code style is that one used in the files. For new files we use [this coy style](https://github.com/ComputationalRadiationPhysics/contributing/blob/master/codingGuideLines/cpp.md)
